File: libs/go/bar.go
package stonecharts

import (
	"fmt"
	"math"
	"strconv"
	"strings"
)

const barPad = 0.2

// renderBarSVG is a one-line delegation to the shared cartesian frame: band
// scale, include-zero value axis, a11y noun "Bar", orientation "horizontal".
// Bar is column with the band axis on y and the value axis on x — see
// charts/bar/design.md for the full geometry contract.
func renderBarSVG(spec *ChartSpec) string {
	return renderCartesian(spec, "Bar", "band", barMarks, true, "horizontal")
}

// barMarks emits the bar-specific marks — one <g class="sc-series"> per series
// and one baseline/floating <rect> per datum, widened along x instead of
// column's height along y. All scales come from the frame; this function owns
// only the pinned band sub-layout and stack transform, transposed.
func barMarks(f *cartesianFrame, p *strings.Builder) {
	if f.n <= 0 {
		return
	}

	bandHeight := f.bandHeight()
	groupH := bandHeight * (1 - barPad)
	stacked := f.stacking == "normal" || f.stacking == "percent"
	kSlots := len(f.spec.Series)
	if stacked || !f.spec.groupingOn() {
		kSlots = 1
	}
	if kSlots <= 0 {
		kSlots = 1
	}
	barH := groupH / float64(kSlots)
	baseline := f.valueZero()

	totals := make([]float64, f.n)
	if stacked {
		for _, s := range f.spec.Series {
			for i, v := range s.Data {
				if i < f.n {
					totals[i] += v
				}
			}
		}
	}

	positive := make([]float64, f.n)
	negative := make([]float64, f.n)
	for si, s := range f.spec.Series {
		st := f.styles[si]
		p.WriteString(fmt.Sprintf(`<g class="sc-series" data-series="%d">`, si))
		for i, raw := range s.Data {
			if i >= f.n {
				break
			}
			cyBand := f.bandCenter(i)
			var top, x, w, tip float64
			if stacked {
				top = cyBand - groupH/2
				value := raw
				if f.stacking == "percent" {
					total := totals[i]
					if total == 0 {
						value = 0
					} else {
						value = raw / total * 100.0
					}
				}
				var leftV, rightV float64
				if value >= 0 {
					leftV = positive[i]
					rightV = leftV + value
					positive[i] = rightV
				} else {
					leftV = negative[i]
					rightV = leftV + value
					negative[i] = rightV
				}
				x0 := f.valuePix(leftV)
				x1 := f.valuePix(rightV)
				x = math.Min(x0, x1)
				w = math.Abs(x0 - x1)
				tip = x1
			} else {
				slot := 0
				if f.spec.groupingOn() {
					slot = si
				}
				top = cyBand - groupH/2 + barH*float64(slot)
				xv := f.valuePix(raw)
				x = math.Min(baseline, xv)
				w = math.Abs(baseline - xv)
				tip = xv
			}
			ylabel := strconv.Itoa(i)
			if i < len(f.cats) {
				ylabel = f.cats[i]
			}
			cy := top + barH/2
			common := fmt.Sprintf(
				`class="sc-bar sc-point" data-series="%d" data-series-name="%s" data-x="%s" data-y="%s" data-color="%s" data-r="3.5" data-r-hover="6"`,
				si, esc(s.Name), esc(ylabel), esc(fmtNum(raw)), st.solid)
			p.WriteString(fmt.Sprintf(
				`<rect %s cx="%s" cy="%s" x="%s" y="%s" width="%s" height="%s" fill="%s"/>`,
				common, f1(tip), f1(cy), f1(x), f1(top), f1(w), f1(barH), st.fill))
		}
		p.WriteString(`</g>`)
	}
}

File: libs/python/stonecharts/charts/bar.py
"""Bar chart renderer: ChartSpec -> SVG string.

Bar is column transposed: the value axis runs along x, the category (band)
axis runs along y. Shared Cartesian chrome (including the orientation-aware
axis/gridline chrome) comes from _cartesian.py; this module draws only bar
marks: one baseline-anchored <rect> per category/series segment, widened
along x instead of column's height along y. See charts/bar/design.md for the
full geometry contract.
"""
from __future__ import annotations

from ..util import esc, fmt_num
from ._cartesian import CartesianFrame, render_cartesian


PAD = 0.2


def render_svg(spec) -> str:
    return render_cartesian(spec, "Bar", "band", _bar_marks, orientation="horizontal")


def _bar_marks(fr: CartesianFrame, p: list) -> None:
    if fr.n <= 0:
        return

    band_height = fr.band_height()
    group_h = band_height * (1 - PAD)
    stacked = fr.stacking in ("normal", "percent")
    k_slots = 1 if stacked or not fr.spec.grouping else max(len(fr.spec.series), 1)
    bar_h = group_h / k_slots
    baseline = fr.value_zero()

    totals = [0.0] * fr.n
    if stacked:
        for s in fr.spec.series:
            for i, v in enumerate(s.data):
                if i < fr.n:
                    totals[i] += v

    positive = [0.0] * fr.n
    negative = [0.0] * fr.n
    for si, s in enumerate(fr.spec.series):
        st = fr.styles[si]
        p.append(f'<g class="sc-series" data-series="{si}">')
        for i, raw in enumerate(s.data):
            if i >= fr.n:
                break
            cy_band = fr.band_center(i)
            if stacked:
                top = cy_band - group_h / 2
                if fr.stacking == "percent":
                    total = totals[i]
                    value = 0.0 if total == 0.0 else raw / total * 100.0
                else:
                    value = raw
                if value >= 0:
                    left_v = positive[i]
                    right_v = left_v + value
                    positive[i] = right_v
                else:
                    left_v = negative[i]
                    right_v = left_v + value
                    negative[i] = right_v
                x0 = fr.value_pix(left_v)
                x1 = fr.value_pix(right_v)
                x = min(x0, x1)
                w = abs(x0 - x1)
                tip = x1
            else:
                top = cy_band - group_h / 2 + bar_h * (si if fr.spec.grouping else 0)
                xv = fr.value_pix(raw)
                x = min(baseline, xv)
                w = abs(baseline - xv)
                tip = xv
            ylabel = fr.cats[i] if i < len(fr.cats) else str(i)
            cy = top + bar_h / 2
            common = (
                f'class="sc-bar sc-point" data-series="{si}" data-series-name="{esc(s.name)}" '
                f'data-x="{esc(ylabel)}" data-y="{esc(fmt_num(raw))}" '
                f'data-color="{st.solid}" data-r="3.5" data-r-hover="6"'
            )
            p.append(
                f'<rect {common} cx="{tip:.1f}" cy="{cy:.1f}" x="{x:.1f}" y="{top:.1f}" '
                f'width="{w:.1f}" height="{bar_h:.1f}" fill="{st.fill}"/>'
            )
        p.append("</g>")
